scivision/catalog/catalog.py

```diff
@@ -1,85 +1,225 @@
+    format: str
+    labels: bool
+    institution: Optional[str]
+    tags: Tuple[str, ...]
+
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+
+class CatalogDatasources(BaseModel, extra="forbid"):
+    catalog_type: str = "scivision datasource catalog"
+    name: str
+    entries: Tuple[CatalogDatasourceEntry, ...]
+
+
+def _coerce_datasources_catalog(
+    datasources: Union[CatalogDatasources, os.PathLike, None]
+) -> CatalogDatasources:
+    """Returns a CatalogDatasources determined from the argument: either
+    the one passed, or one loaded from a file
+    """
+    if isinstance(datasources, CatalogDatasources):
+        return datasources
+    elif isinstance(datasources, (bytes, str, os.PathLike)):
+        datasources_raw = Path(datasources).read_text()
+        return CatalogDatasources.parse_raw(datasources_raw)
+    elif datasources is None:
+        datasources_raw = pkgutil.get_data(__name__, "data/datasources.json")
+        return CatalogDatasources.parse_raw(datasources_raw)
+    else:
+        raise TypeError("Cannot load datasource from unsupported type")
+
+
+def _coerce_models_catalog(
+    models: Union[CatalogModels, os.PathLike, None]
+) -> CatalogModels:
+    """Returns a CatalogModels determined from the argument: either the
+    one passed, or one loaded from a file
+    """
+    if isinstance(models, CatalogModels):
+        return models
+    elif isinstance(models, (bytes, str, os.PathLike)):
+        models_raw = Path(models).read_text()
+        return CatalogModels.parse_raw(models_raw)
+    elif models is None:
+        models_raw = pkgutil.get_data(__name__, "data/models.json")
+        return CatalogModels.parse_raw(models_raw)
+    else:
+        raise TypeError("Cannot load datasource from unsupported type")
+
+
+class PandasCatalog:
+    def __init__(self, datasources=None, models=None):
         super().__init__()
 
-        datasources_uri = pkgutil.get_data(__name__, "data/datasources.json")
-        models_uri = pkgutil.get_data(__name__, "data/models.json")
+        if isinstance(datasources, pd.DataFrame):
+            self._datasources = datasources
+        else:
+            datasources_cat = _coerce_datasources_catalog(datasources)
+            self._datasources = pd.DataFrame(
+                [ent.dict() for ent in datasources_cat.entries]
+            )
 
-        datasources = pd.read_json(datasources_uri, orient="index")
-        models = pd.read_json(models_uri, orient="index")
+        if isinstance(models, pd.DataFrame):
+            self._models = models
+        else:
+            models_cat = _coerce_models_catalog(models)
+            self._models = pd.DataFrame([ent.dict() for ent in models_cat.entries])
 
-        self._models = models.explode("task").explode("data_format")
-        self._datasources = datasources.explode("task").explode("format")
+    @property
+    def models(self) -> pd.DataFrame:
+        return self._models
```

Comment:
Does this bind us to using dataframes as the default type for listing models and datasources? If we change the catalog backend (i.e. not use Pandas for queries etc) we're effectively saying everything downstream will use a Pandas dataframe.

Reply:
This makes sense. To some extent I like the idea of models and datasources themselves being primitive objects (that maybe have a `to_pandas()` method), as it seems like it would give the most flexibility in the long run, by providing a consistent interface, but allowing us to chage the internals. But I realise that's a fair bit of engineering to get something working, epsecially since the dataframe approach works well in practice. Also, pandas is great and everyone uses it.

Reply:
Perhaps now a bit overengineered, but the queries now return instances of `QueryResult`, which can be converted to other types (dict, dataframe).

Reply:
I have to say - I quite like this `QueryResult` now that I see it.  We can add some nice features like an html representation of the result (a la dask arrays, using `__repr_html__`) amongst other things.